sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java

```diff
@@ -0,0 +1,501 @@
+          .build();
+    }
+
+    public DynamoDBConfiguration withRegion(String region) {
+      return withRegion(ValueProvider.StaticValueProvider.of(region));
+    }
+
+    public DynamoDBConfiguration withRegion(ValueProvider<String> region) {
+      return builder().setRegion(region).build();
+    }
+
+    public DynamoDBConfiguration withEndpointUrl(String endpointUrl) {
+      return withEndpointUrl(ValueProvider.StaticValueProvider.of(endpointUrl));
+    }
+
+    public DynamoDBConfiguration withEndpointUrl(ValueProvider<String> endpointUrl) {
+      return builder().setRegion(endpointUrl).build();
+    }
+
+    public DynamoDBConfiguration withAwsAccessKey(String accessKey) {
+      return withAwsAccessKey(ValueProvider.StaticValueProvider.of(accessKey));
+    }
+
+    public DynamoDBConfiguration withAwsAccessKey(ValueProvider<String> accessKey) {
+      return builder().setAwsAccessKey(accessKey).build();
+    }
+
+    public DynamoDBConfiguration withAwsSecretKey(String secretKey) {
+      return withAwsSecretKey(ValueProvider.StaticValueProvider.of(secretKey));
+    }
+
+    public DynamoDBConfiguration withAwsSecretKey(ValueProvider<String> secretKey) {
+      return builder().setAwsSecretKey(secretKey).build();
+    }
+
+    AmazonDynamoDB buildAmazonDynamoDB() {
+      AmazonDynamoDBClientBuilder builder = AmazonDynamoDBClientBuilder.standard();
+      if (getEndpointUrl() != null) {
+        builder.setEndpointConfiguration(
+            new AwsClientBuilder.EndpointConfiguration(getEndpointUrl().get(), getRegion().get()));
+      }
+      if (getEndpointUrl() == null && getRegion() != null) {
+        builder.setRegion(getRegion().get());
+      }
+      if (getAwsAccessKey() != null && getAwsSecretKey() != null) {
+        builder.setCredentials(
+            new AWSStaticCredentialsProvider(
+                new BasicAWSCredentials(getAwsAccessKey().get(), getAwsSecretKey().get())));
+      }
+      return builder.build();
+    }
+  }
+
+  /** Read data from DynamoDB and return PCollection<Map<String, AttributeValue>>. */
+  @AutoValue
+  public abstract static class Read
+      extends PTransform<PBegin, PCollection<Map<String, AttributeValue>>> {
+    @Nullable
+    abstract DynamoDBConfiguration getDynamoDBConfiguration();
+
+    @Nullable
+    abstract SerializableFunction<Void, ScanRequest> getScanRequestFn();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+
+      abstract Builder setDynamoDBConfiguration(DynamoDBConfiguration dynamoDBConfiguration);
+
+      abstract Builder setScanRequestFn(SerializableFunction<Void, ScanRequest> fn);
+
+      abstract Read build();
+    }
+
+    public Read withDynamoDBConfiguration(DynamoDBConfiguration dynamoDBConfiguration) {
+      return toBuilder().setDynamoDBConfiguration(dynamoDBConfiguration).build();
+    }
+
+    public Read withScanRequestFn(SerializableFunction<Void, ScanRequest> fn) {
```

Comment:
Given that ScanRequest is Serializable what is the reason to not pass this directly (I mean without wrapping it in the SerializableFunction). Can you change it to do it directly if there are no constraints to do so?

Reply:
Yes, we can try that. I though you meant to implement the withScanRequestFn in lambda style, so ...

Reply:
Mmm no maybe that was a confusion, the most straight forward the API is the better.

Reply:
@iemejia , have tried sending ScanReqeust directly, but doesn't work. The ScanReqeust seem to be serializable, but it also extends `com.amazonaws.AmazonWebServiceRequest`, which is not serializable, and AmazonWebServiceRequest contains quite a lot of objects/properties that are not serializable. When testing passing ScanRequest directly, it fails most of unit tests that have more than one split (or totalSegment > 1). Any idea of how it could be done beside `SerializableFunction<Void, ScanRequest>` ?

Reply:
No, that was what I was afraid of, ok so let's let it like it is then. Thanks.